WAF Optimize systems - Manage cloud costs and Tag cloud resources (#1203)
hashicorp/web-unified-docs: content/well-architected-framework/docs/docs/optimize-systems/manage-cost/create-cloud-budgets.mdx

Implementing a budget provides you with the following benefits:

- **Visibility into cloud spending:** Understand where your money is going.
- **Proactive cost management:** Take action before costs exceed budgets.
- **Notification of spending anomalies:** Get alerts when spending patterns change.
- **Ability to improve financial planning and forecasting:** Use historical data to make informed budget decisions.

<Note>

The Terraform example in this document has a `tags` block. Refer to the [Tag cloud resources](/well-architected-framework/optimize-systems/lifecycle-management/tag-cloud-resources) document to learn about implementing a tagging strategy.

</Note>

## Create spending limits and notifications

> Most major cloud providers offer native tools to create budgets. These tools allow you to set budget thresholds, monitor spending, and receive alerts when spending approaches or exceeds defined limits. 

Reviewer: ```suggestion
Most major cloud providers offer native tools to create budgets. These native tools allow you to set budget thresholds, monitor spending, and receive alerts when spending approaches or exceeds defined limits. 
```

The following is an example of a Terraform configuration that creates an AWS EC2 budget. This budget tracks EC2 instance costs and sends an alert to test@example.com when the forecasted cost exceeds 100% of the budget. You can set similar budgets and alerts for other cloud providers, such as Azure and GCP.

```hcl
resource "aws_budgets_budget" "ec2" {
  name              = "budget-ec2-monthly"
  budget_type       = "COST"
  limit_amount      = "1200"
  limit_unit        = "USD"
  time_period_end   = "2087-06-15_00:00"
  time_period_start = "2017-07-01_00:00"
  time_unit         = "MONTHLY"

  cost_filter {
    name = "Service"
    values = [
      "Amazon Elastic Compute Cloud - Compute",
    ]
  }

  notification {
    comparison_operator        = "GREATER_THAN"
    threshold                  = 100
    threshold_type             = "PERCENTAGE"
    notification_type          = "FORECASTED"
    subscriber_email_addresses = ["test@example.com"]
  }

  tags = {
    Environment = "production"
    Team        = "engineering"
    ManagedBy   = "terraform"
  }
}
```

Some of the key components in the previous example include:
- **limit_amount:** Defines the monthly spend limit.

> - **notification:** Defines the notification criteria and sends an email to test@example.com.

Reviewer: ```suggestion
- **notification:** Defines the notification criteria, including the recipient email.
```
Nit: The config doesn't _technically_ send the email, just configures it on AWS

HashiCorp resources:

- Learn how to [Tag cloud resources](/well-architected-framework/optimize-systems/lifecycle-management/tag-cloud-resources)
- Terraform resource: [aws_budgets_budget](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/budgets_budget)
- Terraform resource: [azurerm_consumption_budget_subscription](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/consumption_budget_subscription)
- Terraform resource: [google_billing_budget](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/billing_budget)

External resources:

- AWS Budgets: [Getting started with AWS Budgets](https://docs.aws.amazon.com/cost-management/latest/userguide/budgets-managing-costs.html)
- Azure Cost Management and Billing: [Create and manage budgets](https://learn.microsoft.com/en-us/azure/cost-management-billing/costs/tutorial-acm-create-budgets)
- Google Cloud Budgets and alerts: [Creating budgets](https://cloud.google.com/billing/docs/how-to/budgets)

## Detect spending anomalies

> Anomaly detection identifies unusual patterns rather than absolute thresholds. For example, if your monthly EC2 spending suddenly doubles from $2,000 to $4,000 but remains under your $5,000 budget, a budget alert would not trigger. However, anomaly detection would flag this unusual increase for investigation. Anomaly detection helps you catch issues like misconfigured autoscaling, forgotten resources, or unauthorized usage before they significantly impact costs.

Reviewer: ```suggestion
Anomaly detection identifies unusual patterns rather than absolute thresholds. For example, if your monthly EC2 spending suddenly doubles from $2,000 to $4,000 but remains under your $5,000 budget, a budget alert would not trigger. However, anomaly detection would flag this unusual increase for investigation. Anomaly detection helps you identify issues like misconfigured autoscaling, forgotten resources, or unauthorized usage before they significantly impact costs.
```

Most cloud providers offer machine learning-based anomaly detection that learns your normal usage patterns and alerts you when spending deviates from the baseline. You can configure anomaly detection using Terraform for AWS Cost Anomaly Detection and Azure Cost Management.

> The following is an example Terraform code that sets up a cost anomaly detection with email alerts in AWS. This cost anomaly detection will detect the previous EC2 scenario.

Reviewer: ```suggestion
The following is an example Terraform configuration that sets up cost anomaly detection with email alerts in AWS. This cost anomaly detection will detect the previous EC2 scenario.
```
